crates/bevy_transform/src/hierarchy/child_builder.rs

```diff
@@ -11,25 +12,49 @@ pub struct InsertChildren {
 
 impl Command for InsertChildren {
     fn write(self: Box<Self>, world: &mut World, _resources: &mut Resources) {
-        for child in self.children.iter() {
-            world
-                .insert(*child, (Parent(self.parent), PreviousParent(self.parent)))
-                .unwrap();
-        }
-        {
-            let mut added = false;
-            if let Ok(mut children) = world.get_mut::<Children>(self.parent) {
-                children.0.insert_from_slice(self.index, &self.children);
-                added = true;
+        let mut childset = HashSet::default();
+
+        let mut new_children = if let Ok(children) = world.get::<Children>(self.parent) {
+            childset.extend(children.iter().copied());
```

Comment:
I think the cost of this might become significant if the number of children is large and the game is making frequent hierarchy changes. That's why in #1079 I chose to make the set persistent. It makes Children more complex internally but avoids this full iteration of the list on every operation.

Reply:
removed that by now.

Reply:
Have a look again. I now "only" check for contain in the smallvec itself, though for each added child.

I refrain from maintaining a hash map or hash set in the children component itself, since I still need to maintain order with a vec, which defeats the purpose on insert and removal and imposes more allocations for many components with few children.

I refrain from creating an ad-hoc hash set to "speed up" contains checks per operation, since its a rather delicate perf optimization which needs to be benchmarked to do it properly. I could imagine for use cases with many children and many insert/removal operations, an order-independent approach is more suitable. Like going pure HashSet in that case.